nengo/builder/optimizer.py

```diff
@@ -0,0 +1,969 @@
+"""Operator graph optimizers."""
+
+from collections import defaultdict
+import logging
+import time
+import warnings
+
+import numpy as np
+
+from nengo.builder.neurons import SimNeurons
+from nengo.builder import operator
+from nengo.builder.operator import DotInc, ElementwiseInc, SlicedCopy
+from nengo.builder.signal import Signal
+from nengo.config import SupportRcDefaultsMixin
+from nengo.params import BoolParam, IntParam, RcDefault
+from nengo.utils.compat import zip_longest
+from nengo.utils.graphs import toposort, transitive_closure
+
+logger = logging.getLogger(__name__)
+
+
+class OpMergeOptimizer(SupportRcDefaultsMixin):
```

Comment:
I could maybe see separating out some of the code checking for mergeability and dispatching that stuff to the `*Merger` classes. However:
* Parts of that code are inherently tied to the rest of the implementation. For example in `_perform_merges_for_view_subset` it is essential that the comparison in the call to `self._check_sequential(merge[-1], op2)` uses `<` and not `<=`. By separating this in separate classes it might appear that there is less coupling than there actually is.
* I'm not sure what sort of construct would be right. All of the methods I would separate out wouldn't require a class that gets instantiated. Should I define all the methods as `@classmethod`? Or should I make them module methods?